File: code/design-patterns/structural/decorator/go/decorator.go
package main

import (
	"fmt"
)

// Notifier is the interface that defines the send method.
type Notifier interface {
	Send()
}

// SimpleNotifier is a basic implementation of Notifier.
type SimpleNotifier struct{}

func (n *SimpleNotifier) Send() {
	fmt.Println("Basic notification.")
}

// EmailNotifier is a concrete decorator that adds email notification.
type EmailNotifier struct {
	notifier Notifier
}

func (n *EmailNotifier) Send() {
	n.notifier.Send()
	n.sendEmail()
}

func (n *EmailNotifier) sendEmail() {
	fmt.Println("Sending email notification.")
}

// SMSNotifier is a concrete decorator that adds SMS notification.
type SMSNotifier struct {
	notifier Notifier
}

func (n *SMSNotifier) Send() {
	n.notifier.Send()
	n.sendSMS()
}

func (n *SMSNotifier) sendSMS() {
	fmt.Println("Sending SMS notification.")
}

// FacebookNotifier is a concrete decorator that adds Facebook notification.
type FacebookNotifier struct {
	notifier Notifier
}

func (n *FacebookNotifier) Send() {
	n.notifier.Send()
	n.sendFacebook()
}

func (n *FacebookNotifier) sendFacebook() {
	fmt.Println("Sending Facebook notification.")
}

// SlackNotifier is a concrete decorator that adds Slack notification.
type SlackNotifier struct {
	notifier Notifier
}

func (n *SlackNotifier) Send() {
	n.notifier.Send()
	n.sendSlack()
}

func (n *SlackNotifier) sendSlack() {
	fmt.Println("Sending Slack notification.")
}

File: code/design-patterns/structural/decorator/python/decorator.py
from abc import ABC, abstractmethod


class Notifier(ABC):
    @abstractmethod
    def send(self):
        pass


class BaseNotifierDecorator(Notifier):
    def __init__(self, notifier: Notifier):
        self._wrapped_notifier = notifier

    def send(self):
        self._wrapped_notifier.send()


class EmailNotifier(BaseNotifierDecorator):
    def send(self):
        super().send()
        self._send_email()

    def _send_email(self):
        print("Sending email notification.")


class SMSNotifier(BaseNotifierDecorator):
    def send(self):
        super().send()
        self._send_sms()

    def _send_sms(self):
        print("Sending SMS notification.")


class FacebookNotifier(BaseNotifierDecorator):
    def send(self):
        super().send()
        self._send_facebook()

    def _send_facebook(self):
        print("Sending Facebook notification.")


class SlackNotifier(BaseNotifierDecorator):
    def send(self):
        super().send()
        self._send_slack()

    def _send_slack(self):
        print("Sending Slack notification.")


# Example usage
class SimpleNotifier(Notifier):
    def send(self):
        print("Basic notification.")
